xdsl/dialects/emitc.py

```diff
@@ -0,0 +1,87 @@
+"""
+Dialect to generate C/C++ from MLIR.
+
+The EmitC dialect allows to convert operations from other MLIR dialects to EmitC ops.
+Those can be translated to C/C++ via the Cpp emitter.
+
+See external [documentation](https://mlir.llvm.org/docs/Dialects/EmitC/).
+"""
+
+from collections.abc import Iterable
+
+from xdsl.dialects.builtin import (
+    ArrayAttr,
+    ContainerType,
+    IntAttr,
+    ShapedType,
+)
+from xdsl.ir import (
+    AttributeCovT,
+    Dialect,
+    ParametrizedAttribute,
+    TypeAttribute,
+)
+from xdsl.irdl import (
+    ParameterDef,
+    irdl_attr_definition,
+)
+from xdsl.parser import AttrParser
+from xdsl.printer import Printer
+
+
+@irdl_attr_definition
+class EmitC_ArrayType(
+    ParametrizedAttribute, TypeAttribute, ShapedType, ContainerType[AttributeCovT]
+):
+    """EmitC array type"""
+
+    name = "emitc.array"
+
+    shape: ParameterDef[ArrayAttr[IntAttr]]
+    element_type: ParameterDef[AttributeCovT]
+
+    def __init__(
+        self,
+        shape: Iterable[int | IntAttr],
+        element_type: AttributeCovT,
+    ):
+        shape = ArrayAttr(
+            [IntAttr(dim) if isinstance(dim, int) else dim for dim in shape]
+        )
+        if not shape.data:
+            raise VerifyException("EmitC_ArrayType shape must not be empty")
+        super().__init__([shape, element_type])
+
+    def get_num_dims(self) -> int:
+        return len(self.shape.data)
+
+    def get_shape(self) -> tuple[int, ...]:
+        return tuple(i.data for i in self.shape.data)
+
+    def get_element_type(self) -> AttributeCovT:
+        return self.element_type
+
+    @classmethod
+    def parse_parameters(cls, parser: AttrParser):
+        with parser.in_angle_brackets():
+            shape, type = parser.parse_ranked_shape()
+            if not shape:
+                raise parser.raise_error("EmitC_ArrayType shape must not be empty")
```

Comment:
Oh we should have a separate parsing test also from verification, that is done with the `--parser-diagnostics` flag to `xdsl-op`, and we should add a `pytest` test for the verify exception

Reply:
the empty shape verifying is only tested in a pytest now, don't think it is possible from a lit test since that fails already at parsing